### PR TITLE
[Oracle Module] Handle special characters for Oracle connection string properly

### PR DESCRIPTION
**Pull request by yug-elastic**
- Enhancement

## What does this PR do?

This PR enhances the Oracle Metricbeat Module to handle the special characters in the connection string properly.

## Why is it important?

Many of the passwords are machine generated and often contain `%` at the very least, the workaround in the existing module was to not use these characters.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
~~- [ ] I have made corresponding change to the default configuration files~~
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Related issues

- Relates #24609

### CODE REVIEW

**Comment by yug-elastic** [comment]
Hey @cmacknz, all the checks for this PR in the CI are green but E2E tests and packaging parts seem to be failing.
Is this something we can fix as a part of this PR?
Also, the E2E tests are only running for some specific PRs (for example, they do not seem to run for https://github.com/elastic/beats/pull/31462). Is it related to any specific change in this PR?

Please guide us on the same at your convenience. Thanks!

File: x-pack/metricbeat/module/oracle/performance/data.go
Comment (jsoriano) [inline x-pack/metricbeat/module/oracle/performance/data.go:77]: This name extractor wouldn't work with normal urls, that seem to be also supported by `godror.ParseDSN`. Ideally you should be using the host included in the result of `godror.ParseDSN`.

File: x-pack/metricbeat/module/oracle/connection.go
Comment (jsoriano) [inline x-pack/metricbeat/module/oracle/connection.go:46]: Would it be possible to move all this url parsing logic to an url parser? This would be a more natural way to set later the `service.name` for example.

Look for example to the url parser for postgresql: https://github.com/elastic/beats/blob/ac7071bcf36233f216d12dd7473a20aacb6153ea/metricbeat/module/postgresql/postgresql.go#L55

I think that the host set [here](https://github.com/elastic/beats/blob/ac7071bcf36233f216d12dd7473a20aacb6153ea/metricbeat/module/postgresql/postgresql.go#L98) is later used as `service.address`.

Comment (yug-elastic) [reply x-pack/metricbeat/module/oracle/connection.go:46]: Thanks for passing the references. However, this logic is not for URL parsing but is basically for using the `username` and `password` if they are mentioned as the separate fields in the configuration. 

Please refer to the last two fields (commented) in the below configuration:
```
- module: oracle
  period: 10s
  metricsets:
    - performance
  enabled: true
  hosts: ['user="user" password="pass" connectString="0.0.0.0:1521/ORCLPDB1.localdomain"']

  # username: ""
  # password: ""
```

This is the reason why we haven't moved to a URL parser. Also, it would probably make it a bit more complex.
What are your thoughts?

File: x-pack/metricbeat/module/oracle/performance/data.go
Comment (yug-elastic) [reply x-pack/metricbeat/module/oracle/performance/data.go:77]: Yes, we tried the same before and it definitely sounds like a better way. But, the problem with it was that it collected the same value for the field `service.address` when we tried passing multiple hosts in the `oracle.yml` configuration.

When we were digging deeper into it after your comment, we were wondering how the Oracle module iterates through multiple hosts provided in the configuration. What is the reason behind passing just `Hosts[0]` [here](https://github.com/elastic/beats/blob/main/x-pack/metricbeat/module/oracle/connection.go#L40)?

We also tried collecting data from two different Oracle hosts, but most of the metrics looked similar to us.

Was the above problem because of this or are we missing something here? Can you please point us in the right direction?

Comment (jsoriano) [reply x-pack/metricbeat/module/oracle/performance/data.go:77]: > Was the above problem because of this or are we missing something here? Can you please point us in the right direction?

I think this module needs to be refactored to properly use host parsers instead of doing its own parsing of `hosts`. This could be done in a separated PR.

File: x-pack/metricbeat/module/oracle/connection.go
Comment (jsoriano) [reply x-pack/metricbeat/module/oracle/connection.go:46]: > However, this logic is not for URL parsing but is basically for using the `username` and `password` if they are mentioned as the separate fields in the configuration.

Well, it does a bit of both things, it parses the URL, and configures `username` and `password` if provided. Then it generates another URL, stored in `URI`, that can be used by the metricset. Like [here](https://github.com/elastic/beats/blob/ac7071bcf36233f216d12dd7473a20aacb6153ea/metricbeat/module/postgresql/metricset.go#L47) in the postgresql case.

> This is the reason why we haven't moved to a URL parser. Also, it would probably make it a bit more complex.
What are your thoughts?

Not sure if more complex, but using an URL parser has some advantages:
* It is a single place to handle particularities of connection strings for certain services.
* When parsing the configured host, it allows to set in the `HostData` the `Host`, that is later used for `service.address`.
* It abstracts the handling of multiple `hosts`, each metricset instance receives a single `HostData` with the host they have to use.
* Apart of `username` and `password`, it could handle other module-specific settings related to the url/connection string.

I think that using `HostData` is the only way a metricset instance has to know what host it should use, when multiple `hosts` are configured. And the URL parser is the responsible of configuring this `HostData`.

File: x-pack/metricbeat/module/oracle/performance/data.go
Comment (yug-elastic) [reply x-pack/metricbeat/module/oracle/performance/data.go:77]: Thanks a lot for the detailed explanation, @jsoriano! That resolved the doubts and made things pretty clear.
I have raised an issue for the same, and added it to the existing [Oracle Module Enhancements PRD](https://github.com/elastic/beats/issues/30810).

Issue Link: https://github.com/elastic/beats/issues/31611
I have added details to the issue based on our conversations. Please feel free to update the issue with more details if required or in case I missed anything.

File: x-pack/metricbeat/module/oracle/connection.go
Comment (yug-elastic) [reply x-pack/metricbeat/module/oracle/connection.go:46]: Understood, thanks for explaining the advantages! 
I believe we would need to refactor the module as per https://github.com/elastic/beats/issues/31611 in order to move forward with this PR.

File: x-pack/metricbeat/module/oracle/performance/data.go
Comment (yug-elastic) [reply x-pack/metricbeat/module/oracle/performance/data.go:77]: > I think this module needs to be refactored to properly use host parsers instead of doing its own parsing of hosts. This could be done in a separated PR.

Raised the PR https://github.com/elastic/beats/pull/31692 to refactor the module to do the same, please take a look at your convenience.

**Comment by yug-elastic** [comment]
Hey @agithomas, this PR is ready for review. Please ignore the E2E test failures in the CI. We have a [Slack Conversation](https://elastic.slack.com/archives/CJMURHEHX/p1654236314085059) going on in the [observablt-robots](https://elastic.slack.com/archives/CJMURHEHX) channel, @mdelapenya and the team is looking into it and we'll be able to merge this PR once the CI issues are fixed.

Please feel free to provide your feedback over the PR meanwhile.

**Comment by mdelapenya** [comment]
I'd retry the tests today, as we merged the automated bump to level versions to 8.4 in https://github.com/elastic/e2e-testing/pull/2597

**Comment by agithomas** [comment]
Please validate and confirm if the below connection strings are supported. This is present in the original issue

```
 # hosts: ['oracle://c##elkuser:%!@#$^&*()~`<>?[]{}:oracleprd123:1525/PRD_DATA']
  # hosts: ["oracle://c##elkuser:\!@#$^&*()~`<>?[]{}:oracleprd123:1525/PRD_DATA"]
  # hosts: ['oracle://c##elkuser:\!@#$^&*()~`<>?[]{}:oracleprd123:1525/PRD_DATA']
  
  # Works without % or \
  # hosts: ["oracle://c##elkuser:!@#$^&*()~`<>?[]{}:oracleprd123:1525/PRD_DATA"]
  # hosts: ['oracle://c##elkuser:!@#$^&*()~`<>?[]{}:oracleprd123:1525/PRD_DATA']
```

File: metricbeat/docs/modules/oracle.asciidoc
Comment (agithomas) [inline metricbeat/docs/modules/oracle.asciidoc:88]: What i understand is 

The old format is hosts: ["user/pass@0.0.0.0:1521/ORCLPDB1.localdomain"]

New format is 

hosts: ['user="user" password="pass" connectString="0.0.0.0:1521/ORCLPDB1.localdomain"']

Is this a breaking change then ?

Comment (yug-elastic) [reply metricbeat/docs/modules/oracle.asciidoc:88]: Your understanding is correct. We went with using the DSN format to configure the module, as discussed with @jsoriano over the TDD. I am not sure if we should consider this as a breaking change or an enhancement, please let me know if I should change the position of the changelog entry accordingly.

Comment (agithomas) [reply metricbeat/docs/modules/oracle.asciidoc:88]: With this new format, the existing users of sql module for oracle might be impacted. With this logic, to me, it is a breaking change. Let us discuss in the team meeting.

**Comment by yug-elastic** [comment]
> Please validate and confirm if the below connection strings are supported. This is present in the original issue
> 
> ```
> hosts: ["oracle://c##elkuser:%!@#$^&*()~`<>?[]{}:oracleprd123:1525/PRD_DATA"]
>  # hosts: ['oracle://c##elkuser:%!@#$^&*()~`<>?[]{}:oracleprd123:1525/PRD_DATA']
>   # hosts: ["oracle://c##elkuser:\!@#$^&*()~`<>?[]{}:oracleprd123:1525/PRD_DATA"]
>   # hosts: ['oracle://c##elkuser:\!@#$^&*()~`<>?[]{}:oracleprd123:1525/PRD_DATA']
>   
>   # Works without % or \
>   # hosts: ["oracle://c##elkuser:!@#$^&*()~`<>?[]{}:oracleprd123:1525/PRD_DATA"]
>   # hosts: ['oracle://c##elkuser:!@#$^&*()~`<>?[]{}:oracleprd123:1525/PRD_DATA']
> ```

Yes, we have confirmed that all the connection strings mentioned are supported. One thing to note here, connection string with `\` is supported but the user will need to escape the `\` i.e. if the password is `\!@#$^&*()~` as mentioned in the 3rd example, in the configuration, the user will have to mention the password as `\\!@#$^&*()~`. This has also been discussed over the TDD as this is how "yaml.v2" package processes the configuration. Backslashes in strings between double quotes are used as escape characters in YAML.

File: metricbeat/docs/modules/oracle.asciidoc
Comment (jsoriano) [reply metricbeat/docs/modules/oracle.asciidoc:88]: I think we should continue supporting both formats if possible, even if it is only for backwards compatibility reasons. This is something we do in the postgresql module for example.

**Review by kush-elastic — APPROVED** [review]
LGTM!

File: metricbeat/docs/modules/oracle.asciidoc
Comment (agithomas) [reply metricbeat/docs/modules/oracle.asciidoc:88]: I agree @yug-elastic . It won't be a breaking change if it is backward compatible .

Comment (agithomas) [reply metricbeat/docs/modules/oracle.asciidoc:88]: Kindly update the documentation to mention the support of both formats.

Also, please update the need of escaping the backslash character.

Comment (yug-elastic) [reply metricbeat/docs/modules/oracle.asciidoc:100]: Makes sense, thank you.
Updated the documentation ([link](https://github.com/elastic/beats/pull/31368/files#diff-bb326e2221aee29daf0c82e51a6006849f5f691086a48b26e97640696dbda392)). Please take a look.

**Review by agithomas — APPROVED** [review]
LGTM

**Comment by yug-elastic** [comment]
> @yug-elastic I'm adding a 👍 in terms of the e2e tests as, for this PR, they have suffered a glitch in the parallel stages that they run. Because the vast majority of the tests are passing, and after a deeper investigation on the existing failures, I'd say that they are temporary and not related to this PR

Thank you for the approval, @mdelapenya! I wonder if it would be possible to merge this PR until we have a green checkmark over `beats-ci/pr-merge` check. If it's possible, kindly help us with it as we don't have the authority to merge this PR.

FYI, we have an approval from @agithomas who is the code owner of this change.

File: x-pack/metricbeat/module/oracle/_meta/docs.asciidoc
Comment (agithomas) [inline x-pack/metricbeat/module/oracle/_meta/docs.asciidoc:55]: Please include the below mentioned configuration as well

user="user" password="password" connectString="host:port/service_name" sysdba=true

and

"user/password@0.0.0.0:1521/ORCLPDB1.localdomain as sysdba"

These are a bit tricky configs and let us document this to address questions that may arise in futre

Comment (yug-elastic) [reply x-pack/metricbeat/module/oracle/_meta/docs.asciidoc]: Sure, updated the documentation. Thanks!

**Comment by cmacknz** [comment]
The E2E tests here are being triggered because of the updates to go.mod. There's nothing there that looks like it should be causing test failures, and I've seen the same set of tests failing on other branches recently, like https://beats-ci.elastic.co/job/e2e-tests/job/e2e-testing-mbp/job/main/1256/ for one example.

I'm going to merge this.

**Comment by yug-elastic** [comment]
Thanks a lot, @cmacknz!

**Pull request merged**
